File: src/scala/axis-cli/src/main/scala/command/Universe.scala
package command

import picocli.CommandLine.{Command, HelpCommand, Option, Parameters}
import data.{AxisUniverse, DataUtils}

@Command(name="universe", subcommands = Array(classOf[HelpCommand]))
class Universe {
    @Command
    def create(@Parameters name: String, @Option(names=Array("--description","-d"), defaultValue = "") description: String): Unit = {
        AxisUniverse.create(name, description)
    }

    @Command
    def list(): Unit = {
        var universes = AxisUniverse.list()
        if (universes.nonEmpty) {
            DataUtils.printNonEmptyItems(universes.collect(r => r.toMap()), AxisUniverse.printFields)
        }
    }

    @Command
    def show(@Parameters name: String): Unit = {
        var universes = AxisUniverse.find(name)
        DataUtils.printNonEmptyItems(universes.collect(r => r.toMap()), AxisUniverse.printFields)
    }

    @Command
    def delete(@Parameters name: String,  @Option(names=Array("--force","-f"), defaultValue = "false") force : Boolean): Unit = {
        AxisUniverse.delete(name)
    }

    @Command
    def update(@Parameters name: String, @Option(names = Array("--name")) newName: String, @Option(names = Array("--description", "-d")) newDescription: String): Unit = {
        AxisUniverse.update(name, newName, newDescription)
    }
}

File: src/swift/axis-cli/Sources/axis-cli/command/Universe.swift
import ArgumentParser



struct Universe: ParsableCommand {
    static var configuration = CommandConfiguration(
            subcommands: [Create.self, List.self, Show.self, Delete.self, Update.self])
}


extension Universe {
    struct Create: ParsableCommand {
        @Argument var name: String
        @Option(name: .shortAndLong) var description: String?

        mutating func run() throws {
            try AxisUniverse.Create(name: name, description: description)
        }
    }

    struct List: ParsableCommand {
        mutating func run() throws {
            let universes = try AxisUniverse.List()
            AxisUniverse.printRecords(universes: universes)
        }
    }

    struct Show: ParsableCommand {
        @Argument var name: String

        mutating func run() throws {
            let universe = try AxisUniverse.Find(name: name)
            AxisUniverse.printRecords(universes: [universe])
        }
    }

    struct Delete: ParsableCommand {
        @Argument var name: String

        mutating func run() throws {
            try AxisUniverse.Delete(name: name)
        }
    }

    struct Update: ParsableCommand {
        @Argument var name: String
        @Option(name: .customLong("name")) var newName: String?
        @Option(name: .shortAndLong) var description: String?

        mutating func run() throws {
            try AxisUniverse.Update(name: name, newName: newName, description: description)
        }
    }
}